pkg/skaffold/docker/dependencies.go

```diff
@@ -283,7 +283,26 @@ func WalkWorkspace(workspace string, excludes, deps []string) (map[string]bool,
 		}); err != nil {
 			return nil, fmt.Errorf("walking %q: %w", absFrom, err)
 		}
-	}
 
+		// Check if directory is already in the files map
+		if err := walk.From(absFrom).Unsorted().WhenIsDir().When(keepFile).Do(func(path string, info walk.Dirent) error {
+			relPath, err := filepath.Rel(workspace, path)
+			if err != nil {
+				return err
+			}
+			var keepDir = true
+			for item := range files {
+				if strings.Contains(item, relPath) {
+					keepDir = false
+				}
+			}
```

Comment:
This will be very expensive and also suffers from false positives.  Consider the following:
```
test/  <-- empty directory
testing/README.txt
src/main/test/Test.java
```

`testing/README.txt` contains `test`
`src/main/test/Main.java` contains `test`

There are a few approaches I can think of here:
1. Change the original walk to process both files and directories.  If the path is a directory, check if the directory has contents, and if empty then add the directory to `files`.  This could entail adding a `walk.IsEmpty(dir string) bool` funciton, whose implementation would use [godirwalk's NewScanner](https://github.com/karrick/godirwalk/blob/f64d91544257ab0cc1cca44f22600925a1b9ddae/examples/remove-empty-directories/main.go#L50-L66).
2. Do two passes, where the first pass collects the parent directories of files to a separate `directoriesWithChildren` map. Then walk the directories in the second pass and only add directories to `files` if they weren't in `directoriesWithChildren`.
3. Add a new `DoThen(Action, postDir Action)` action to our `walk` interface that adds a post-directory action using [`godirwalk`'s PostChildrenCallback](https://github.com/karrick/godirwalk/tree/f64d91544257ab0cc1cca44f22600925a1b9ddae#configurable-post-children-callback).  Then do the same `directoriesWithChildren` building as in Option 2 but then check empty-dirs in this post-children callback.

Reply:
Thank you very much for your comment. I agree with you about the extra cost of going over all the files for each directory (although preventing false positives should be easy to do). 

My initial idea was using a function to check if the directory is empty (so basically the first approach you described). That would avoid having to make two passes. 

I'm going to have a look at the godirwalk to see how it's being done. I'll let you know if I get stuck.